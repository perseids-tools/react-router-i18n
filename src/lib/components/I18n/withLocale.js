import React from 'react';
import PropTypes from 'prop-types';

import { withRouter } from 'react-router-dom';

const withLocale = (WrappedComponent) => {
  const NewComponent = ({
    to, match, ignoreLocale, staticContext, ...rest
  }) => {
    const { params } = match;

    if (ignoreLocale) {
      return (
        <WrappedComponent
          to={to}
          match={match}
          {...rest}
        />
      );
    }

    let path = to;

    if (params && params.locale) {
      path = path[0] === '/'
        ? `/${params.locale}${path}`
        : `${params.locale}/${path}`;
    }

    return (
      <WrappedComponent
        to={path}
        match={match}
        {...rest}
      />
    );
  };

  NewComponent.propTypes = {
    to: PropTypes.string.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        locale: PropTypes.string,
      }),
    }).isRequired,
    ignoreLocale: PropTypes.bool,
    // eslint-disable-next-line react/forbid-prop-types
    staticContext: PropTypes.object,
  };

  NewComponent.defaultProps = {
    ignoreLocale: false,
    staticContext: undefined,
  };

  return withRouter(NewComponent);
};

export default withLocale;
