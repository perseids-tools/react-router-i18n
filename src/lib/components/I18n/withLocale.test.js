import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter, Route } from 'react-router-dom';
import withLocale from './withLocale';

const OriginalComponent = ({ to, text }) => (
  <div>
    {to}
    {text}
  </div>
);
const WrappedComponent = withLocale(OriginalComponent);

it('renders without crashing', () => {
  const component = (
    <MemoryRouter>
      <WrappedComponent to="test" />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders normally when there is no match', () => {
  const component = (
    <MemoryRouter initialEntries={['/home/']}>
      <Route exact path="/:locale?/home" component={() => <WrappedComponent to="test" />} />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('adds the locale to `to` when there is a match', () => {
  const component = (
    <MemoryRouter initialEntries={['/fr/home/']}>
      <Route exact path="/:locale?/home" component={() => <WrappedComponent to="test" />} />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('passes through any other arguments', () => {
  const component = (
    <MemoryRouter>
      <WrappedComponent to="test" text="example" />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
