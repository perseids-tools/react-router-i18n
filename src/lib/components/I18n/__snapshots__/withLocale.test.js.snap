// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds the locale to \`to\` when there is a match 1`] = `
<div>
  fr/test
</div>
`;

exports[`passes through any other arguments 1`] = `
<div>
  test
  example
</div>
`;

exports[`renders normally when there is no match 1`] = `
<div>
  test
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  test
</div>
`;
